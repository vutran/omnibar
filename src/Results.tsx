import * as React from 'react';
import ResultsItem from './ResultsItem';
import { COLORS } from './constants';

interface Props<T> {
  // the currently selected index
  selectedIndex: number;
  // list of result items
  items: Array<T>;
  // onMouseEnter callback
  onMouseEnter?: (e: any /* Event */) => void;
  // onMouseLeave callback
  onMouseLeave?: (e: any /* Event */) => void;
  // onMouseEnter item callback
  onMouseEnterItem?: (e: any /* Event */) => void;
  // onMouseLeave item callback
  onMouseLeaveItem?: (e: any /* Event */) => void;
  // onClick callback
  onClickItem?: (e: any /* Event */) => void;
  // max container height
  maxHeight?: React.CSSLength;
  // item row height
  rowHeight?: React.CSSLength;
  // optional override container style
  style?: React.CSSProperties;
  // optional row override style
  rowStyle?: React.CSSProperties;
  // optional result renderering function
  resultRenderer?: <T>(item: T) => React.ReactChild;
}

const LIST_STYLE: React.CSSProperties = {
  position: 'absolute',
  width: '100%',
  zIndex: 2,
  listStyleType: 'none',
  margin: 0,
  padding: 0,
  backgroundColor: COLORS.WHITE,
};

export default function Results<T>(props: Props<T>) {
  const style = { ...LIST_STYLE, ...props.style };

  if (props.maxHeight) {
    style.maxHeight = props.maxHeight;
    style.borderBottomWidth = 1;
    style.borderBottomColor = COLORS.GRAY;
    style.borderBottomStyle = 'solid';
    style.overflow = 'auto';
  }

  function createHandler(handler: any, key: number) {
    return handler.bind(this, key);
  }

  return (
    <ul
      style={style}
      onMouseEnter={props.onMouseEnter}
      onMouseLeave={props.onMouseLeave}
    >
      {props.items.map((item, key) =>
        React.createElement(ResultsItem, {
          key,
          highlighted: props.selectedIndex === key,
          item,
          style: props.rowStyle,
          onMouseEnter:
            props.onMouseEnterItem &&
            createHandler(props.onMouseEnterItem, key),
          onClickItem: props.onClickItem,
          resultRenderer: props.resultRenderer,
        })
      )}
    </ul>
  );
}
