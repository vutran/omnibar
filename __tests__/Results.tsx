import * as React from 'react';
import Results from '../src/Results';
import renderer from 'react-test-renderer';

describe('Results', () => {
  const items = [
    { title: 'Google', url: 'https://google.com' },
    { title: 'Dropbox', url: 'https://dropbox.com' },
  ];

  it('should render a <Results />', () => {
    const tree = renderer
      .create(
        React.createElement(Results, {
          items,
          selectedIndex: -1,
        })
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a <Results /> with a selected item', () => {
    const tree = renderer
      .create(
        React.createElement(Results, {
          items,
          selectedIndex: 0,
        })
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a <Results /> with a max height', () => {
    const tree = renderer
      .create(
        React.createElement(Results, {
          items,
          selectedIndex: -1,
          maxHeight: 200,
        })
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a <Results /> with a row height', () => {
    const tree = renderer
      .create(
        React.createElement(Results, {
          items,
          selectedIndex: -1,
          maxHeight: 200,
        })
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a <Results /> with a custom row renderer', () => {
    const rowRenderer = ({ item }: { item: any }) => (
      <div>Title: {item.title}</div>
    );
    const tree = renderer
      .create(
        React.createElement(Results, {
          items,
          selectedIndex: -1,
          resultRenderer: rowRenderer,
        })
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
