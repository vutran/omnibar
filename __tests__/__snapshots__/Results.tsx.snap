// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results should render a <Results /> 1`] = `
<ul
  onMouseEnter={undefined}
  onMouseLeave={undefined}
  style={
    Object {
      "backgroundColor": "#fff",
      "listStyleType": "none",
      "margin": 0,
      "padding": 0,
      "position": "absolute",
      "width": "100%",
      "zIndex": 2,
    }
  }
>
  <a
    href="https://google.com"
    item={
      Object {
        "title": "Google",
        "url": "https://google.com",
      }
    }
    onClick={undefined}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#ddd",
        "borderLeftWidth": 1,
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopWidth": 0,
        "boxSizing": "border-box",
        "color": "#000",
        "display": "block",
        "fontSize": 24,
        "height": 50,
        "lineHeight": "50px",
        "paddingLeft": 15,
        "paddingRight": 15,
        "textDecoration": "none",
      }
    }
  >
    Google
  </a>
  <a
    href="https://dropbox.com"
    item={
      Object {
        "title": "Dropbox",
        "url": "https://dropbox.com",
      }
    }
    onClick={undefined}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#ddd",
        "borderLeftWidth": 1,
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopWidth": 0,
        "boxSizing": "border-box",
        "color": "#000",
        "display": "block",
        "fontSize": 24,
        "height": 50,
        "lineHeight": "50px",
        "paddingLeft": 15,
        "paddingRight": 15,
        "textDecoration": "none",
      }
    }
  >
    Dropbox
  </a>
</ul>
`;

exports[`Results should render a <Results /> with a custom row renderer 1`] = `
<ul
  onMouseEnter={undefined}
  onMouseLeave={undefined}
  style={
    Object {
      "backgroundColor": "#fff",
      "listStyleType": "none",
      "margin": 0,
      "padding": 0,
      "position": "absolute",
      "width": "100%",
      "zIndex": 2,
    }
  }
>
  <div>
    Title: 
    Google
  </div>
  <div>
    Title: 
    Dropbox
  </div>
</ul>
`;

exports[`Results should render a <Results /> with a max height 1`] = `
<ul
  onMouseEnter={undefined}
  onMouseLeave={undefined}
  style={
    Object {
      "backgroundColor": "#fff",
      "borderBottomColor": "#eee",
      "borderBottomStyle": "solid",
      "borderBottomWidth": 1,
      "listStyleType": "none",
      "margin": 0,
      "maxHeight": 200,
      "overflow": "auto",
      "padding": 0,
      "position": "absolute",
      "width": "100%",
      "zIndex": 2,
    }
  }
>
  <a
    href="https://google.com"
    item={
      Object {
        "title": "Google",
        "url": "https://google.com",
      }
    }
    onClick={undefined}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#ddd",
        "borderLeftWidth": 1,
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopWidth": 0,
        "boxSizing": "border-box",
        "color": "#000",
        "display": "block",
        "fontSize": 24,
        "height": 50,
        "lineHeight": "50px",
        "paddingLeft": 15,
        "paddingRight": 15,
        "textDecoration": "none",
      }
    }
  >
    Google
  </a>
  <a
    href="https://dropbox.com"
    item={
      Object {
        "title": "Dropbox",
        "url": "https://dropbox.com",
      }
    }
    onClick={undefined}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#ddd",
        "borderLeftWidth": 1,
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopWidth": 0,
        "boxSizing": "border-box",
        "color": "#000",
        "display": "block",
        "fontSize": 24,
        "height": 50,
        "lineHeight": "50px",
        "paddingLeft": 15,
        "paddingRight": 15,
        "textDecoration": "none",
      }
    }
  >
    Dropbox
  </a>
</ul>
`;

exports[`Results should render a <Results /> with a row height 1`] = `
<ul
  onMouseEnter={undefined}
  onMouseLeave={undefined}
  style={
    Object {
      "backgroundColor": "#fff",
      "borderBottomColor": "#eee",
      "borderBottomStyle": "solid",
      "borderBottomWidth": 1,
      "listStyleType": "none",
      "margin": 0,
      "maxHeight": 200,
      "overflow": "auto",
      "padding": 0,
      "position": "absolute",
      "width": "100%",
      "zIndex": 2,
    }
  }
>
  <a
    href="https://google.com"
    item={
      Object {
        "title": "Google",
        "url": "https://google.com",
      }
    }
    onClick={undefined}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#ddd",
        "borderLeftWidth": 1,
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopWidth": 0,
        "boxSizing": "border-box",
        "color": "#000",
        "display": "block",
        "fontSize": 24,
        "height": 50,
        "lineHeight": "50px",
        "paddingLeft": 15,
        "paddingRight": 15,
        "textDecoration": "none",
      }
    }
  >
    Google
  </a>
  <a
    href="https://dropbox.com"
    item={
      Object {
        "title": "Dropbox",
        "url": "https://dropbox.com",
      }
    }
    onClick={undefined}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#ddd",
        "borderLeftWidth": 1,
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopWidth": 0,
        "boxSizing": "border-box",
        "color": "#000",
        "display": "block",
        "fontSize": 24,
        "height": 50,
        "lineHeight": "50px",
        "paddingLeft": 15,
        "paddingRight": 15,
        "textDecoration": "none",
      }
    }
  >
    Dropbox
  </a>
</ul>
`;

exports[`Results should render a <Results /> with a selected item 1`] = `
<ul
  onMouseEnter={undefined}
  onMouseLeave={undefined}
  style={
    Object {
      "backgroundColor": "#fff",
      "listStyleType": "none",
      "margin": 0,
      "padding": 0,
      "position": "absolute",
      "width": "100%",
      "zIndex": 2,
    }
  }
>
  <a
    href="https://google.com"
    item={
      Object {
        "title": "Google",
        "url": "https://google.com",
      }
    }
    onClick={undefined}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "backgroundColor": "#eee",
        "borderBottomWidth": 1,
        "borderColor": "#ddd",
        "borderLeftWidth": 1,
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopWidth": 0,
        "boxSizing": "border-box",
        "color": "#000",
        "display": "block",
        "fontSize": 24,
        "height": 50,
        "lineHeight": "50px",
        "paddingLeft": 15,
        "paddingRight": 15,
        "textDecoration": "none",
      }
    }
  >
    Google
  </a>
  <a
    href="https://dropbox.com"
    item={
      Object {
        "title": "Dropbox",
        "url": "https://dropbox.com",
      }
    }
    onClick={undefined}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#ddd",
        "borderLeftWidth": 1,
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopWidth": 0,
        "boxSizing": "border-box",
        "color": "#000",
        "display": "block",
        "fontSize": 24,
        "height": 50,
        "lineHeight": "50px",
        "paddingLeft": 15,
        "paddingRight": 15,
        "textDecoration": "none",
      }
    }
  >
    Dropbox
  </a>
</ul>
`;
